src/target/riscv/riscv-013_reg.c

```diff
@@ -13,73 +13,452 @@
+	}
+	int res = access_prelude(reg, /*write*/ true);
+	if (res != ERROR_OK)
+		return res;
+	if (reg->valid && buf_eq(reg->value, buf, reg->size)) {
+		LOG_TARGET_DEBUG(riscv_reg_impl_get_target(reg),
+				"Writing the same value to %s.", reg->name);
 	} else {
-		const riscv_reg_t value = buf_get_u64(buf, 0, reg->size);
-		if (riscv_reg_set(target, reg->number, value) != ERROR_OK)
-			return ERROR_FAIL;
+		LOG_TARGET_DEBUG(target, "Caching the write to %s.", reg->name);
+		set_cache_value(reg, buf);
+		mark_dirty(reg);
 	}
+	access_epilogue(reg, /*write*/ true);
+	return ERROR_OK;
+}
 
+static int flush_reg(struct reg *reg, reg_sender send_reg)
+{
+	if (!reg->dirty)
+		return ERROR_OK;
+	int res = access_prelude(reg, /*write*/ true);
+	if (res != ERROR_OK)
+		return res;
+	assert(reg->dirty);
+	assert(reg->valid);
+	const struct target * const target = riscv_reg_impl_get_target(reg);
+	if (target->state != TARGET_HALTED) {
+		LOG_TARGET_ERROR(target,
+				"BUG: register %s is dirty while the target is not halted.",
+				reg->name);
+		return ERROR_TARGET_NOT_HALTED;
+	}
+	res = send_reg(reg, reg->value);
+	if (res != ERROR_OK) {
+		/* Register is not marked "invalid" here (like it's done on failure in
+		 * "caching_set") since it is "dirty", i.e. the most recent value
+		 * is currently in the cache. */
+		return res;
+	}
+	mark_clean(reg);
+	access_epilogue(reg, /*write*/ true);
+	return res;
+}
+
+/* End of cache entry utils. */
+/* Start of "reg_arch_type" method definitions. */
+
+static int riscv013_noncaching_get(struct reg *reg)
+{
+	return noncaching_get(reg, fetch_riscv013_reg);
+}
+
+static int riscv013_caching_get(struct reg *reg)
+{
+	return caching_get(reg, fetch_riscv013_reg);
+}
+
+static int riscv013_noncaching_set(struct reg *reg, uint8_t *buf)
+{
+	return noncaching_set(reg, buf, send_riscv013_reg);
+}
+
+static int riscv013_caching_set(struct reg *reg, uint8_t *buf)
+{
+	return caching_set(reg, buf, send_riscv013_reg);
+}
+
+static int riscv013_reg_flush(struct reg *reg)
+{
+	return flush_reg(reg, send_riscv013_reg);
+}
+
+static int flush_uncacheable(struct reg *reg)
+{
+	assert(!reg->dirty);
+	return ERROR_OK;
+}
+
+static int get_vreg(struct reg *reg)
+{
+	assert(reg->number >= GDB_REGNO_V0 && reg->number <= GDB_REGNO_V31);
+	return caching_get(reg, fetch_vreg);
+}
+
+static int set_vreg(struct reg *reg, uint8_t *buf)
+{
+	assert(reg->number >= GDB_REGNO_V0 && reg->number <= GDB_REGNO_V31);
+	return caching_set(reg, buf, send_vreg);
+}
+
+static int flush_vreg(struct reg *reg)
+{
+	assert(reg->number >= GDB_REGNO_V0 && reg->number <= GDB_REGNO_V31);
+	return flush_reg(reg, send_vreg);
+}
+
+static int get_pc(struct reg *pc)
+{
+	assert(pc->number == GDB_REGNO_PC);
+	assert(!pc->valid);
+	assert(!pc->dirty);
+	struct reg * const dpc =
+		riscv_reg_impl_cache_entry(riscv_reg_impl_get_target(pc),
+				GDB_REGNO_DPC);
+	int res = dpc->type->get(dpc);
+	if (res == ERROR_OK)
+		set_cache_value(pc, dpc->value);
```

Comment:
Did I get the meaning of this operation right?

```suggestion
	if (res == ERROR_OK)
		/* There are two aliases for DPC in the register cache - "pc" and "dpc".
		 * Keep them in sync. */
		set_cache_value(pc, dpc->value);
```

Reply:
Yeah, I'll add the comment.